ext/dom/element.c

```diff
@@ -1246,4 +1246,118 @@ PHP_METHOD(DOMElement, replaceWith)
 }
 /* }}} end DOMElement::prepend */
 
+#define INSERT_ADJACENT_RES_FAILED ((void*) -1)
+
+static xmlNodePtr dom_insert_adjacent(const zend_string *where, xmlNodePtr thisp, dom_object *this_intern, xmlNodePtr otherp)
+{
+	if (zend_string_equals_literal_ci(where, "beforebegin")) {
+		if (thisp->parent == NULL) {
+			return NULL;
+		}
+		if (dom_hierarchy(thisp->parent, otherp) == FAILURE) {
+			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		otherp = xmlAddPrevSibling(thisp, otherp);
+	} else if (zend_string_equals_literal_ci(where, "afterbegin")) {
+		if (dom_hierarchy(thisp, otherp) == FAILURE) {
+			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		if (thisp->children == NULL) {
+			otherp = xmlAddChild(thisp, otherp);
+		} else {
+			otherp = xmlAddPrevSibling(thisp->children, otherp);
+		}
+	} else if (zend_string_equals_literal_ci(where, "beforeend")) {
+		if (dom_hierarchy(thisp, otherp) == FAILURE) {
+			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		otherp = xmlAddChild(thisp, otherp);
+	} else if (zend_string_equals_literal_ci(where, "afterend")) {
+		if (thisp->parent == NULL) {
+			return NULL;
+		}
+		if (dom_hierarchy(thisp->parent, otherp) == FAILURE) {
+			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
+			return INSERT_ADJACENT_RES_FAILED;
+		}
+		otherp = xmlAddNextSibling(thisp, otherp);
+	} else {
+		php_dom_throw_error(SYNTAX_ERR, dom_get_strict_error(this_intern->document));
+		return INSERT_ADJACENT_RES_FAILED;
```

Comment:
The spec seem to indicate that a DOMException *must* be thrown:
 ```suggestion
		php_dom_throw_error(SYNTAX_ERR, true);
		return INSERT_ADJACENT_RES_FAILED;
```

Reply:
Yes, but strangely PHP offers an "stricterror" option (default true) which allows you to get warnings instead of exceptions.
And that maybe doesn't even violate spec, because from https://webidl.spec.whatwg.org/#dfn-throw

> The resulting behavior from creating and throwing an exception is language binding specific.

Reply:
See https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#Document3-strictErrorChecking for more info

Reply:
Yeah PHP extension used to have the trifecto of no warning/exception, warnings, exceptions. I think mainly because prior to the introduction of exception warnings were the only way.

We probably should get rid of those and only allow silent or exception behaviour (similarly to the SQLite3 RFC) but that's future scope